### ISSUE TITLE
'Use of scanf() operation that permits buffer overflows (CWE-120)' issue in src\server\scripts\Commands\cs_titles.cpp

### ISSUE THREAD

**Issue by Kitzunu** [azerothcore/azerothcore-wotlk#5297]
[CodeFactor](https://www.codefactor.io/repository/github/azerothcore/azerothcore-wotlk) found an issue: Use of scanf() operation that permits buffer overflows (CWE-120)

It's currently on:
[src\server\scripts\Commands\cs_titles.cpp:197](https://www.codefactor.io/repository/github/azerothcore/azerothcore-wotlk/source/master/src/server/scripts/Commands/cs_titles.cpp#L197)
Commit f6a17164be8700d4e457a51071ca4034ecbc5b73

**Comment by Kitzunu** [comment]
Is this something we should fix or overlook?

**Comment by alchem1ster** [comment]
First of all, it must be said that any code weakness must be eliminated. But here we are talking exclusively about an internal mechanism, on which the client cannot interact. Thus, if you are sure that `args` will not contain such amount of data that may overflow `uint64` - you can ignore it.
_Source: personal experience and graduate degree in information security_

**Comment by Kitzunu** [comment]
If you feel like you have the time, feel free to check https://www.codefactor.io/repository/github/azerothcore/azerothcore-wotlk/issues?category=Security as there are a dozen security issues. Perhaps some of them are not important, but as I am a beginner at C++ and definitely this stuff it would be nice to have an experienced eye =)

**Comment by alchem1ster** [comment]
@Kitzunu 

I've already looked into it :)
What I would recommend for `CWE-120` and `CWE-20`  is to use the analogs of the features with enhanced security.
For example:
AVOID | USE | ALT USE
------------ | ------------- | -------------
strcat | strlcat | strcat_s
strcpy | strlcpy | strcpy_s
sscanf | sscanf_s |
sprintf | snprintf | sprintf_s
vsprintf | vsnprintf | vspintf_s

And for `CWE-134` you need to specify the format string in `vsprintf` as a constant instead of referring to external entities.

All of these recommendations cover the full scope of identified flaws according to CodeFactor.

**Comment by Kitzunu** [comment]
no longer an issue

**Comment by alchem1ster** [comment]
https://www.codefactor.io/repository/github/azerothcore/azerothcore-wotlk/issues?category=Security still contain some CWE-120
Maybe just swap sprintf to snprintf or sprintf_s? CWE-134 can be skipped if `format` param always constant.
@Kitzunu

**Comment by Kitzunu** [comment]
You are right, I just checked for the this specific file. Will update the issue accordingly

**Comment by Kitzunu** [comment]
Would something like this be the correct way of going about it?

```diff
diff --git a/src/common/Debugging/WheatyExceptionReport.cpp b/src/common/Debugging/WheatyExceptionReport.cpp
index be56630025..76b13ee7b4 100644
--- a/src/common/Debugging/WheatyExceptionReport.cpp
+++ b/src/common/Debugging/WheatyExceptionReport.cpp
@@ -1615,7 +1615,7 @@ int __cdecl WheatyExceptionReport::StackLog(const TCHAR* format, va_list argptr)
     DWORD cbWritten;
 
     TCHAR szBuff[WER_LARGE_BUFFER_SIZE];
-    retValue = vsprintf(szBuff, format, argptr);
+    retValue = _vsnprintf(szBuff, WER_LARGE_BUFFER_SIZE, format, argptr);
     WriteFile(m_hReportFile, szBuff, retValue * sizeof(TCHAR), &cbWritten, 0);
 
     return retValue;
@@ -1628,7 +1628,7 @@ int __cdecl WheatyExceptionReport::HeapLog(const TCHAR* format, va_list argptr)
     TCHAR* szBuff = (TCHAR*)malloc(sizeof(TCHAR) * WER_LARGE_BUFFER_SIZE);
     if (szBuff != nullptr)
     {
-        retValue = vsprintf(szBuff, format, argptr);
+        retValue = _vsnprintf(szBuff, WER_LARGE_BUFFER_SIZE, format, argptr);
         WriteFile(m_hReportFile, szBuff, retValue * sizeof(TCHAR), &cbWritten, 0);
         free(szBuff);
     }
```

**Comment by alchem1ster** [comment]
Sure, but make sure that `WER_LARGE_BUFFER_SIZE` is defined correctly and is of sufficient size for the expected data. This is key, as too small a size can result in data clipping.

**Comment by Kitzunu** [comment]
This is way beyond my comprehension of C++ 😅

Could we do something like
```cpp
if (retValue > WER_LARGE_BUFFER_SIZE)
    retValue = 0
```

**Comment by alchem1ster** [comment]
> Could we do something like

No. If `retValue = 0`, then `WriteFile()` will not be able to write because the `retValue` param is the length of the string that will be written to the file.

WER_LARGE_BUFFER_SIZE = 16384

So strings longer than 16383 (1 for null terminator) characters will be truncated. I think this is more than enough

**Comment by alchem1ster** [comment]
> Would something like this be the correct way of going about it?
> 
> ```diff
> diff --git a/src/common/Debugging/WheatyExceptionReport.cpp b/src/common/Debugging/WheatyExceptionReport.cpp
> index be56630025..76b13ee7b4 100644
> --- a/src/common/Debugging/WheatyExceptionReport.cpp
> +++ b/src/common/Debugging/WheatyExceptionReport.cpp
> @@ -1615,7 +1615,7 @@ int __cdecl WheatyExceptionReport::StackLog(const TCHAR* format, va_list argptr)
>      DWORD cbWritten;
>  
>      TCHAR szBuff[WER_LARGE_BUFFER_SIZE];
> -    retValue = vsprintf(szBuff, format, argptr);
> +    retValue = _vsnprintf(szBuff, WER_LARGE_BUFFER_SIZE, format, argptr);
>      WriteFile(m_hReportFile, szBuff, retValue * sizeof(TCHAR), &cbWritten, 0);
>  
>      return retValue;
> @@ -1628,7 +1628,7 @@ int __cdecl WheatyExceptionReport::HeapLog(const TCHAR* format, va_list argptr)
>      TCHAR* szBuff = (TCHAR*)malloc(sizeof(TCHAR) * WER_LARGE_BUFFER_SIZE);
>      if (szBuff != nullptr)
>      {
> -        retValue = vsprintf(szBuff, format, argptr);
> +        retValue = _vsnprintf(szBuff, WER_LARGE_BUFFER_SIZE, format, argptr);
>          WriteFile(m_hReportFile, szBuff, retValue * sizeof(TCHAR), &cbWritten, 0);
>          free(szBuff);
>      }
> ```

Use this. Should be sufficient.

**Comment by alchem1ster** [comment]
@Kitzunu any progress?

**Comment by Kitzunu** [comment]
I have a lot of rl going on right now so not really prioritizing this atm
